crates/module-system/sov-solana-offchain-auth/src/authentication.rs

```diff
@@ -0,0 +1,483 @@
+use std::fmt::Debug;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use sov_modules_api::capabilities::{
+    calculate_hash_metered, extract_authorization_data, verify_chain_id, AuthenticationError,
+    AuthenticationOutput, FatalError, UniquenessData,
+};
+use sov_modules_api::macros::UniversalWallet;
+use sov_modules_api::transaction::{
+    self, AuthenticatedTransactionAndRawHash, TransactionCallable, TxDetails, UnsignedTransaction,
+};
+use sov_modules_api::{
+    CryptoSpec, DispatchCall, GasMeter, GasSpec, MeteredJsonDeserialize,
+    MeteredJsonDeserializeError, MeteredSignature, ProvableStateReader, Spec, TxHash,
+};
+
+/// The application domain for Solana offchain messages (placeholder for now)
+pub const APPLICATION_DOMAIN: [u8; 32] = [0u8; 32];
+
+/// The payload for a solana offchain message.
+/// Essentially a wrapper around `sov_modules_api::transaction::UnsignedTransaction` that also
+/// includes the chain_hash, in order to ensure the hash gets signed as part of the message.
+/// We duplicate the UnsignedTransaction type rather than wrapping it to ensure the JSON displayed
+/// to the user doesn't get too nested.
+#[serde_with::serde_as]
+#[derive(Debug, Serialize, Deserialize, UniversalWallet)]
+#[serde(bound = "R::Call: serde::Serialize + serde::de::DeserializeOwned")]
+pub struct SolanaOffchainUnsignedTransaction<R: TransactionCallable, S: Spec> {
+    /// The runtime call
+    pub runtime_call: R::Call,
+    /// The uniqueness identifier
+    pub uniqueness: UniquenessData,
+    /// Data related to fees and gas handling.
+    pub details: TxDetails<S>,
+    /// The chain hash the transaction data was signed with. Because we're not actually using the
+    /// schema to display the transaction, we have to include and validate the hash explicitly.
+    #[serde_as(as = "serde_with::hex::Hex")]
+    pub chain_hash: [u8; 32],
+}
+
+impl<R, S> SolanaOffchainUnsignedTransaction<R, S>
+where
+    S: Spec,
+    R: TransactionCallable,
+    <R as TransactionCallable>::Call: Serialize + DeserializeOwned,
+{
+    fn into_unsigned_tx(self) -> UnsignedTransaction<R, S> {
+        UnsignedTransaction {
+            runtime_call: self.runtime_call,
+            uniqueness: self.uniqueness,
+            details: self.details,
+        }
+    }
+
+    fn unmetered_deserialize_inner(buf: &mut &[u8]) -> Result<Self, serde_json::Error> {
+        serde_json::from_slice::<SolanaOffchainUnsignedTransaction<R, S>>(buf)
+    }
+}
+
+impl<R, S> MeteredJsonDeserialize<S> for SolanaOffchainUnsignedTransaction<R, S>
```

Comment:
Why do we need the `MeteredJsonDeserialize` trait? The only usage I’ve seen is
`<SolanaOffchainUnsignedTransaction<D, S> as MeteredJsonDeserialize<S>>::deserialize`,
so it seems like we could just use the `SolanaOffchainUnsignedTransaction` type directly.

Reply:
Simplified as per discussion on slack